File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int n = friends.length;
        Map<String, Integer> friendsIdx = new HashMap<>();
        for (int i = 0; i < n; i++) {
            friendsIdx.put(friends[i], i);
        }
        
        int[][] giftsLog = new int[n][n];
        for (String gift : gifts) {
            String[] temp = gift.split(" ");
            giftsLog[friendsIdx.get(temp[0])][friendsIdx.get(temp[1])] += 1;
        }
        
        int[] giftsIndex = new int[n];
        for (int i = 0; i < n; i++) {
            int give = 0; int take = 0;
            for (int j = 0; j < n; j++) {
                give += giftsLog[i][j];
                take += giftsLog[j][i];
            }
            giftsIndex[i] = give - take;
        }
        
        int[] count = new int[n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (giftsLog[i][j] < giftsLog[j][i]) continue;
                
                if (giftsLog[i][j] > giftsLog[j][i]) {
                    count[i]++;
                } else {
                    if (giftsIndex[i] > giftsIndex[j]) {
                        count[i]++;
                    }
                }
            }
        }
        
        int answer = 0;
        for (int c : count) {
            answer = Math.max(answer, c);
        }
        return answer;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    n = len(friends)
    friends_idx = {friends[i] : i for i, friend in enumerate(friends)}

    gifts_log = [[0] * n for _ in range(n)]
    for gift in gifts:
        a, b = gift.split()
        a_idx, b_idx = friends_idx[a], friends_idx[b]
        gifts_log[a_idx][b_idx] += 1
    
    gifts_index = [0] * n
    for i in range(n):
        give = sum(gifts_log[i])
        take = sum([gifts_log[j][i] for j in range(n)])
        gifts_index[i] = give - take 
    
    answer = [0] * n
    for i in range(n):
        for j in range(n):
            if gifts_log[i][j] < gifts_log[j][i]:
                continue
                
            if gifts_log[i][j] > gifts_log[j][i]:
                answer[i] += 1
            else:
                if gifts_index[i] > gifts_index[j]:
                    answer[i] += 1
    return max(answer)